src/core/geocoding/qgsnominatimgeocoder.cpp

```diff
@@ -0,0 +1,250 @@
+qint64 QgsNominatimGeocoder::sLastRequestTimstamp = 0;
+
+QgsNominatimGeocoder::QgsNominatimGeocoder( const QString &countryCodes, const QString &endpoint )
+  : QgsGeocoderInterface()
+  , mCountryCodes( countryCodes )
+  , mEndpoint( QStringLiteral( "https://nominatim.openstreetmap.org/search" ) )
+{
+  if ( !endpoint.isEmpty() )
+    mEndpoint = endpoint;
+}
+
+QgsGeocoderInterface::Flags QgsNominatimGeocoder::flags() const
+{
+  return QgsGeocoderInterface::Flag::GeocodesStrings;
+}
+
+QgsFields QgsNominatimGeocoder::appendedFields() const
+{
+  QgsFields fields;
+  fields.append( QgsField( QStringLiteral( "osm_type" ), QVariant::String ) );
+  fields.append( QgsField( QStringLiteral( "display_name" ), QVariant::String ) );
+  fields.append( QgsField( QStringLiteral( "place_id" ), QVariant::String ) );
+  fields.append( QgsField( QStringLiteral( "class" ), QVariant::String ) );
+  fields.append( QgsField( QStringLiteral( "type" ), QVariant::String ) );
+  fields.append( QgsField( QStringLiteral( "road" ), QVariant::String ) );
+  fields.append( QgsField( QStringLiteral( "village" ), QVariant::String ) );
+  fields.append( QgsField( QStringLiteral( "city_district" ), QVariant::String ) );
+  fields.append( QgsField( QStringLiteral( "town" ), QVariant::String ) );
+  fields.append( QgsField( QStringLiteral( "city" ), QVariant::String ) );
+  fields.append( QgsField( QStringLiteral( "state" ), QVariant::String ) );
+  fields.append( QgsField( QStringLiteral( "country" ), QVariant::String ) );
+  fields.append( QgsField( QStringLiteral( "postcode" ), QVariant::String ) );
+  return fields;
+}
+
+QgsWkbTypes::Type QgsNominatimGeocoder::wkbType() const
+{
+  return QgsWkbTypes::Point;
+}
+
+QList<QgsGeocoderResult> QgsNominatimGeocoder::geocodeString( const QString &string, const QgsGeocoderContext &context, QgsFeedback *feedback ) const
+{
+  QgsRectangle bounds;
+  if ( !context.areaOfInterest().isEmpty() )
+  {
+    QgsGeometry g = context.areaOfInterest();
+    QgsCoordinateTransform ct( context.areaOfInterestCrs(), QgsCoordinateReferenceSystem( QStringLiteral( "EPSG:4326" ) ), context.transformContext() );
+    try
+    {
+      g.transform( ct );
+      bounds = g.boundingBox();
+    }
+    catch ( QgsCsException & )
+    {
+      QgsDebugMsg( "Could not transform geocode bounds to WGS84" );
+    }
+  }
+
+  const QUrl url = requestUrl( string, bounds );
+
+  QgsReadWriteLocker locker( sMutex, QgsReadWriteLocker::Read );
+  auto it = sCachedResults.constFind( url );
+  if ( it != sCachedResults.constEnd() )
+  {
+    return *it;
+  }
+
+  while ( QDateTime::currentMSecsSinceEpoch() - sLastRequestTimstamp < 1000 / mRequestsPerSecond )
+  {
+    QThread::msleep( 50 );
+    if ( feedback && feedback->isCanceled() )
+      return QList<QgsGeocoderResult>();
+  }
+  locker.changeMode( QgsReadWriteLocker::Write );
+  sLastRequestTimstamp = QDateTime::currentMSecsSinceEpoch();
+  locker.unlock();
```

Comment:
I would move this till after the blocking request call. Then you'd be excluding the length of time it takes to do the actual request from the "1 second between requests" logic and are being a bit more gentle on the service. By leaving the locker.changeMode to the Write mode before the request we ensure that only one request will be happening at a time.

Reply:
@nyalldawson , the way I read Nominatum's usage policy, they set minimum to one request per second, not a 1 second gap between requests. If we move it below the blocking request is done, on very slow connections (i.e., developing parts of the world, where I code from :wink: ), you'd end up with say a request that lasts 1.5sec, then a 1sec gap for a total of 2.5sec.

Reply:
I don't personally see an issue with that, and think we're better to err on the side of being good citizens and keeping the load as light as possible...